man/man1/glnemo.1
.TH GLNEMO 1NEMO "November 2007"
.SH NAME
glnemo \- Interactive 3D OpenGL visualization program for Nemo snapshots (deprecated)
.SH SYNOPSIS
\fBglnemo\fP [parameter=value]
.SH DESCRIPTION
\fBglnemo\fP is an interactive visualization program for snapshots,
based on the Qt library from Trolltech. 
You can zoom in/out, rotate, scale, translate, save a screenshot, 
select different particles
and plot them in different blending colors, display velocity vectors
and make animations . 
.PP
\fBglnemo\fP is depracated, \fBglnemo2\fP has replaced it .
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP
Nemo input snapshot [] or ascii file with a list of NEMO snapshots. In
such case, the ascii file must begin with the following line:
#glnemo_list_file
.nf
example:
-------- cut here --------
#glnemo_list_file
my_snap.0001.nemo 0:199999
my_snap.0002.nemo 100:299999
my_snap.0003.nemo 0:1999,3000:5000
-------- cut here --------
.fi
.TP 20
\fBserver=\fP
Running simulation server hostname (right now works with gyrfalcON) []   
.TP 20
\fBselect=\fP
Select particles. This option can be used to color the particles. Sets of particles
must be selected using the : range operator, separated by a comma. E.g.
0:9,10:19 would select two sets of 10 particles and give them a different color.
[all] 
.TP 20
\fBrange_visib=\fP
Toggle visibility for the particles selected via 'select='
options. Can be usefull if you only want to display particles selected
via 'select_list' options(see below). In that case you should set
'f'.[t]
.TP 20
\fBselect_list=\fP
Select particles from file(s) of list of indexes. You can give several
files separated with ','. Each files must have the following
structure:
.nf
-------- cut here --------
#p_index_list <--- MUST be the first line (file header)
index
index
 ...
 ...
index
-------- cut here --------
.fi
Each indexes must be lower than the #particles read.
.TP 20
\fBtimes=\fP
Select time [all]     
.TP 20
\fBvel=\fP
load velocity coordinates [t]
.TP 20
\fBdisp_vel=\fP
display velocity vectors [f]
.TP 20
\fBblending=\fP
Activate blending colors [t]    
.TP 20
\fBdbuffer=\fP
Activate OpenGL depth buffer [f]
.TP 20
\fBperspective=\fP
Perspective projection is activated by default. Select 'f' to toggle to
orthographic projection.[t]
.TP 20
\fBbestzoom=\fP
If activated, it performs an automatic zoom to fit all particles on the
display.[t]
.TP 20
\fBplay=\fP          
automatically load and display next snapshot [f]
.TP 20
\fBortho_range=\fP
[xy] range size in case of orthographic projection.[6.0]
.TP 20
\fBzoom=\fP
Zoom value.[-14.0]
.TP 20
\fBxrot=\fP
Rotation angle on X axis.[0.0]
.TP 20
\fByrot=\fP
Rotation angle on Y axis.[0.0]
.TP 20
\fBzrot=\fP
Rotation angle on Z axis.[0.0]
.TP 20
\fBxtrans=\fP
translation on X.[0.0]
.TP 20
\fBytrans=\fP
translation on Y.[0.0]
.TP 20
\fBztrans=\fP
translation on Z.[0.0]
.TP 20
\fBgrid=\fP
Show grid [f]
.TP 20
\fBgas=\fP
toggle gas like particles effect.[f]
.TP 20
\fBtexture_s=\fP
texture size for gas like particles, 0 < value <= 1.[0.15]
.TP 20
\fBtexture_ac=\fP
texture alpha color for gas like particles, 0 <= value <= 255.[125]
.TP 20
\fBpsize=\fP
Set particles size [1.0]    
.TP 20
\fBport=\fP
Running simulation server's communication port [4444]    
.TP 20
\fBwsize=\fP
Windows's width size [925]    
.TP 20
\fBhsize=\fP
Windows's height size [685]    
.TP 20
\fBscreenshot=\fP
Screenshot name []
.TP 20
\fBanim_file=\fP
Animation filename (previously recorded from animation module) []
.TP 20
\fBanim_play=\fP
play an animation ? (anim_file must be selected) [t]
.TP 20
\fBanim_bench=\fP
play an animation in benchmark mode ? [t]
.SH OPERATION
All the parameters given to the command line can be interactively
modified in real time using the graphic user interface composed of menus, icons and
dialog boxes. Most of the operations can also be activated by
hitting a key in the display window itself:
.nf
+                               zoom in
-                               zoom out
Wheel Mouse                     zoom in or out
CTRL + Left-Mouse/Right-Mouse   translate
g                               toggle the grid
r                               launch particles ranges/color dialog box
o                               launch option dialog box
CTRL + r                        reset rotation/translation
CTRL + a                        autoscale to fit the frame
s                               take a screenshot
p                               play all snapshots
ALT + g                         toogle gas like particles effect
CTRL + p                        launch printer dialog box
f                               toggle fullscreen (might make it slow)
shift + M                       launch animation box

.fi
.SH SEE ALSO
xyzview(1NEMO), glnemo2(1NEMO)
.PP
http://www.trolltech.com       Qt library web site
.SH LICENCE
CeCILL (see http://www.cecill.info)
.SH FILES
src/nbody/glnemo: directory with source code
.TP
\fBhttp://www.oamp.fr/lam/equipes/dynamique/jcl/glnemo\fP Home page
.SH AUTHOR
Jean-Charles Lambert
.SH UPDATE HISTORY
.nf
.ta +1.0i +5.0i
03-dec-04	V0.50   added to NEMO	JCL
22-jun-05	V0.87   lot of fixes and enhancement	JCL
25-Apr-06	V0.90   animation module, octree, bugs fixed	JCL
15-Jul-06	V0.92   bigger DOF, cube display, bugs fixed	JCL
26-Sep-06	V0.93   velocity vectors display, bugs fixed	JCL
19-Jun-07	V0.94   list of snapshots as input        	JCL
	        snapshots with different #bodies
	        animation files etc...see Changelog
13-Jul-07	V0.94.1 bugs fixed	JCL
22-Nov-07	V0.94.2 bugs fixed on zoom and octree	JCL
.fi


src/nbody/glnemo/src/glnemo.cpp
// ============================================================================
// Copyright Jean-Charles LAMBERT - 2004-2007                                  
// e-mail:   Jean-Charles.Lambert@oamp.fr                                      
// address:  Dynamique des galaxies                                            
//           Laboratoire d'Astrophysique de Marseille                          
//           2, place Le Verrier                                               
//           13248 Marseille Cedex 4, France                                   
//           CNRS U.M.R 6110                                                   
// ============================================================================
//                                                                             
// Glnemo is  an interactive 3D plotting program  , using OpenGL and trolltech 
// QT API. It allows to plot in 3D the positions x,y,z of a NEMO snapshot.     
//                                                                             
// This software is governed by the CeCILL  license under French law and       
// abiding by the rules of distribution of free software.  You can  use,       
// modify and/ or redistribute the software under the terms of the CeCILL      
// license as circulated by CEA, CNRS and INRIA at the following URL           
// "http://www.cecill.info".                                                   
//                                                                             
// As a counterpart to the access to the source code and  rights to copy,      
// modify and redistribute granted by the license, users are provided only     
// with a limited warranty  and the software's author,  the holder of the      
// economic rights,  and the successive licensors  have only  limited          
// liability.                                                                  
//                                                                             
// In this respect, the user's attention is drawn to the risks associated      
// with loading,  using,  modifying and/or developing or reproducing the       
// software by the user in light of its specific status of free software,      
// that may mean  that it is complicated to manipulate,  and  that  also       
// therefore means  that it is reserved for developers  and  experienced       
// professionals having in-depth computer knowledge. Users are therefore       
// encouraged to load and test the software's suitability as regards their     
// requirements in conditions enabling the security of their systems and/or    
// data to be ensured and,  more generally, to use and operate it in the       
// same conditions as regards security.                                        
//                                                                             
// The fact that you are presently reading this means that you have had        
// knowledge of the CeCILL license and that you accept its terms.              
// ============================================================================
//                                                                             
// main program                                                                
//                                                                             
// ============================================================================

#include <iostream>
#include <qapplication.h>
#include <qgl.h>

// Nemo stuffs
#define _vectmath_h // put this statement to avoid conflict with C++ vector class
#include <nemo.h>

#include "globjwin.h"

using namespace std;
// ============================================================================
// NEMO parameters                                                             
const char * defv[] = {  // use `::'string because of 'using namespace std'
  (char *) "in=\n             Nemo input snapshot                                           \n"
           "                   or ascii file with a list of NEMO files                       \n"
           "                   in such case, the first line of the file                    \n"
           "                   must be :                                                   \n"
           "                   #glnemo_file_list                                             ",
  (char *) "server=\n         Running simulation server hostname                              ",
  (char *) "select=all\n      Select particles from the : range operator , separeted        \n"
           "                   by a comma. E.g 0:999,1000:1999 would select two sets        \n"
            "                   of 1000 particles and give them a different color              ",
  (char *) "range_visib=t\n   toggle visibility  for the particles selected via \'select=\' \n"
           "                   options. Can be usefull if you only want to display particles\n"
           "                   selected via \'select_list\' options(see below). In that case\n"
           "                   you should set \'f\'                                           ",
  (char *) "select_list=\n    Select particles from list of indexes                           ",
  (char *) "times=all\n       Select time                                                     ",
  (char *) "vel=t\n           load velocity coordinates                                       ",
  (char *) "disp_vel=f\n      display velocity vectors                                        ",
  (char *) "blending=t\n      Activate blending colors                                        ",
  (char *) "dbuffer=f\n       Activate OpenGL depth buffer                                    ",
  (char *) "perspective=t\n   false means orthographic                                        ",
  (char *) "bestzoom=t\n      automatic zoom                                                  ",
  (char *) "play=f\n          automatically load and display next snapshot                    ",
  (char *) "ortho_range=6.0\n xy range if orthographic projection                             ",
  (char *) "zoom=-14\n        zoom value                                                      ",
  (char *) "xrot=0.0\n        rotation angle on X axis                                        ",
  (char *) "yrot=0.0\n        rotation angle on Y axis                                        ",
  (char *) "zrot=0.0\n        rotation angle on Z axis                                        ",
  (char *) "xtrans=0.0\n      translation on X                                                ",
  (char *) "ytrans=0.0\n      translation on Y                                                ",
  (char *) "ztrans=0.0\n      translation on Z                                                ",
  (char *) "grid=f\n          Show grid                                                       ",
  (char *) "gas=f\n           gas like particle effect                                        ",
  (char *) "texture_s=0.15\n  texture size of gaz particle                                    ",
  (char *) "texture_ac=125\n  texture alpha color of gaz particle                             ",
  (char *) "psize=1.0\n       Set particles size                                              ",
  (char *) "port=4444\n       Server's communication port                                     ",
  (char *) "wsize=925\n       Windows's width size                                            ",
  (char *) "hsize=685\n       Windows's height size                                           ",
  (char *) "screenshot=\n     Screenshot name                                                 ",
  (char *) "anim_file=\n      Animation filename                                              ",
  (char *) "anim_play=t\n     play an animation ? (anim_file must be selected)                ",
  (char *) "anim_bench=t\n    play an animation in benchmark mode ?                           ",
  (char *) "VERSION=0.94.3\n    "__DATE__"  - JCL  compiled at <"__TIME__">                     ",
  NULL
};
const char * usage=(char *) "Interactive 3D OpenGL visualization program for Nemo snapshots";

// ============================================================================
//  The main program is here                                                   
int main( int argc, char **argv )
{
  QApplication::setColorSpec( QApplication::CustomColor );
  QApplication a(argc,argv);			

  if ( !QGLFormat::hasOpenGL() ) {
    qWarning( "This system has no OpenGL support. Exiting." );
    return -1;
  }
  // Get screen coordinates
  QDesktopWidget  * desktop = QApplication::desktop();
  
  // initialyze NEMO engine
  initparam(argv,const_cast<char**>(defv));

  // CAUTION !!! do not call getparam function after **GLObjectWindow** object
  // instantiation bc nemo engine get corrupted by io_nemo afterwhile the     
  // snapshot has been loaded.                                                
  const int wsize=getiparam((char *) "wsize");
  const int hsize=getiparam((char *) "hsize");
  ::string screenshot=NULL;
  bool has_screenshot=false;
  if ( hasvalue((char *) "screenshot")) {
    screenshot=getparam((char *) "screenshot");
    has_screenshot=true;
  }
  // Create widget window
  GLObjectWindow w(0,"glnemo");

  // move window to the center of the screen
  w.setGeometry(desktop->width() /2 - (wsize/2),
                desktop->height()/2 - (hsize/2),
                wsize,hsize);

  a.setMainWidget( &w );
  w.show();
  
  w.resize(wsize-1,hsize-1); // trick to update GL buffer with right W & H
  w.glbox->enable_s=true;
  if (has_screenshot) {
    w.showMinimized();
    w.takeScreenshot(screenshot);
  }
  finiparam();  // garbage collecting for nemo
  if (! has_screenshot) {
    return a.exec();
  }
}
// ============================================================================
